File: app/src/main/java/pers/victor/androiddemos/view/CircleRevealView.java
package pers.victor.androiddemos.view;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Path;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.ImageView;

/**
 * Created by Victor on 2017/3/5. (ง •̀_•́)ง
 */

public class CircleRevealView extends ImageView {
    private Path circlePath = new Path();
    private float radius = 0f;
    private ValueAnimator anim = new ValueAnimator();

    public CircleRevealView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int w = MeasureSpec.getSize(widthMeasureSpec);
        int h = MeasureSpec.getSize(heightMeasureSpec);
        anim.setFloatValues(0f, w > h ? w : h);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        circlePath.reset();
        circlePath.addCircle((getWidth() / 2), (getHeight() / 2), radius, Path.Direction.CW);
        canvas.save();
        canvas.clipPath(circlePath);
        super.onDraw(canvas);
        canvas.restore();
    }

    private void init() {
        anim.setInterpolator(new AccelerateDecelerateInterpolator());
        anim.setDuration(1500);
        anim.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                radius = (float) animation.getAnimatedValue();
                postInvalidate();
            }
        });
    }

    public void circleReveal() {
        if (!anim.isRunning()) {
            anim.start();
        }
    }
}

File: app/src/main/kotlin/demos/view/CircleRevealView.kt
package demos.view

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Path
import android.util.AttributeSet
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.ImageView

/**
 * Author : victor
 * Time : 16-9-17 16:39
 */
class CircleRevealView(context: Context, attributeSet: AttributeSet) : ImageView(context, attributeSet) {
    private val circlePath = Path()
    private var radius = 0f
    private var maxRadius = 0f
    private val anim = ValueAnimator()

    init {
        anim.interpolator = AccelerateDecelerateInterpolator()
        anim.duration = 1500
        anim.addUpdateListener {
            radius = it.animatedValue as Float
            invalidate()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val w = MeasureSpec.getSize(widthMeasureSpec)
        val h = MeasureSpec.getSize(heightMeasureSpec)
        maxRadius = (if (w > h) w else h).toFloat()
        anim.setFloatValues(0f, maxRadius)
    }

    override fun onDraw(canvas: Canvas) {
        circlePath.reset()
        circlePath.addCircle((width / 2).toFloat(), (height / 2).toFloat(), radius, Path.Direction.CW)
        canvas.save()
        canvas.clipPath(circlePath)
        super.onDraw(canvas)
        canvas.restore()
    }

    fun circleReveal() {
        if (!anim.isRunning) {
            anim.start()
        }
    }
}